unix,stream: clear read/write states on close

vtjnash opened the pull request:
Rebased copy of #1825 with some minor style cleanup

saghul commented on src/unix/stream.c:
```diff
@@ -1001,6 +1001,7 @@ uv_handle_type uv__handle_type(int fd) {
 static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
   stream->flags |= UV_HANDLE_READ_EOF;
   stream->flags &= ~UV_HANDLE_READING;
+  stream->flags &= ~UV_HANDLE_READABLE;
```
how about writable?

vtjnash replied:
Does reaching the end of the input make a stream immediately non-writable? That feels false (e.g. what about `shutdown()`), although I often do end up coupling them too.

saghul commented on test/test-not-readable-nor-writable-on-read-error.c:
```diff
@@ -0,0 +1,104 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
libuv contributors

vtjnash replied:
`Copyright the libuv project contributors. All rights reserved.` ? That's used on 5 other files. All of the rest have this header.

bnoordhuis replied:
Yes, that's the one. New files get the new copyright boilerplate.

saghul commented on test/echo-server.c:
```diff
@@ -68,6 +68,24 @@ static void after_shutdown(uv_shutdown_t* req, int status) {
 
+static void dont_linger(uv_handle_t* handle) {
+  struct linger so_linger;
+  uv_os_fd_t socket_fd = 0;
```
initialize to -1 for safety?

saghul commented on test/echo-server.c:
```diff
@@ -68,6 +68,24 @@ static void after_shutdown(uv_shutdown_t* req, int status) {
+
+  so_linger.l_onoff = 1;
+  so_linger.l_linger = 0;
+  ASSERT(0 == setsockopt((uv_os_sock_t) socket_fd,
```
no problems with this on Windows?

vtjnash replied:
This capability is now equivalent to `uv_tcp_close_reset`, so I suppose it should be fine. I've refactored it to use `uv_tcp_close_reset` directly now.

saghul commented on test/test-not-readable-nor-writable-on-read-error.c:
```diff
@@ -0,0 +1,104 @@
+  reset_me = uv_buf_init(reset_me_cmd, sizeof(reset_me_cmd));
+
+  r = uv_write(&write_req,
+                      (uv_stream_t*) &tcp_client,
```
style: align arguments

saghul commented on test/test-not-readable-on-eof.c:
```diff
@@ -0,0 +1,103 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
libuv contributors

saghul commented on test/test-not-readable-on-eof.c:
```diff
@@ -0,0 +1,103 @@
+  read_cb_called++;
+
+  ASSERT(nread == UV_EOF);
+  ASSERT(0 == uv_is_readable(handle));
```
check also if it's writable?

saghul commented on test/test-not-readable-on-eof.c:
```diff
@@ -0,0 +1,103 @@
+  close_me = uv_buf_init(close_me_cmd, sizeof(close_me_cmd));
+
+  r = uv_write(&write_req,
+                      (uv_stream_t*) &tcp_client,
```
style: align args

saghul commented on test/test-not-writable-after-shutdown.c:
```diff
@@ -0,0 +1,69 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
ditto

saghul commented on test/test.gyp:
```diff
@@ -162,6 +162,9 @@
         'test-udp-multicast-interface6.c',
         'test-udp-try-send.c',
         'test-uname.c',
+        'test-not-writable-after-shutdown.c',
```
sort alphabetically

vtjnash replied:
Is is worthwhile to sort the whole list or to make a separate PR for doing that?

saghul commented on test/test.gyp:
```diff
@@ -162,6 +162,9 @@
         'test-uname.c',
+        'test-not-writable-after-shutdown.c',
+        'test-not-readable-nor-writable-on-read-error.c',
+        'test-not-readable-on-eof.c'
```
we need these in the CMakeLists too

bnoordhuis commented on test/test-not-readable-nor-writable-on-read-error.c:
```diff
@@ -0,0 +1,104 @@
+uv_tcp_t tcp_client;
+uv_connect_t connect_req;
+uv_write_t write_req;
+char reset_me_cmd[3] = {'Q', 'S', 'H'};
```
These should all be `static`.

bnoordhuis commented on test/test-not-readable-on-eof.c:
```diff
@@ -0,0 +1,103 @@
+uv_tcp_t tcp_client;
+uv_connect_t connect_req;
+uv_write_t write_req;
+char close_me_cmd[2] = {'Q', 'S'};
```
Ditto.

vtjnash commented:
https://ci.nodejs.org/job/libuv-test-commit/1568/

bnoordhuis commented on test/test.gyp:
```diff
@@ -10,85 +10,92 @@
         'test-multiple-listen.c',
+        'test-mutexes.c',
+        'test-not-readable-nor-writable-on-read-error.c',
+        'test-not-readable-on-eof.c'
```
An interesting factoid about gyp files: if you omit the comma, it concatenates the strings. ^^

(The build failures are caused by the missing comma here.)

vtjnash commented:
CI: https://ci.nodejs.org/job/libuv-test-commit/1673/

@saghul did you want to re-review? I'm assuming I should just go ahead and land this, since I think I addressed all comments (and @bnoordhuis subsequently approved).

vtjnash commented:
It appears that the new `uv_tcp_close_reset` function may not work on AIX:

```
not ok 161 - not_readable_nor_writable_on_read_error
# exit code 393222
# Output from process `not_readable_nor_writable_on_read_error`: (no output)
Output from process `not_readable_nor_writable_on_read_error`:
# Assertion failed in test/test-not-readable-nor-writable-on-read-error.c on line 56: (nread < 0) && (nread != UV_EOF)
```

vtjnash commented:
Ah, digging further, I see that AIX is now (un)intentionally supposed to fail to handle this error condition (https://github.com/libuv/libuv/pull/2447). If I revert that PR, AIX starts emitting errors correctly again (8ab6a79 / https://ci.nodejs.org/job/libuv-test-commit-aix/1717/). I'll disable this test on AIX and open an issue.

vtjnash commented:
new CI: https://ci.nodejs.org/job/libuv-test-commit/1679/

vtjnash commented:
Node CI: https://ci.nodejs.org/view/libuv/job/libuv-in-node/

vtjnash commented:
Can someone with knowledge of nodejs tests help me interpret that result? I see several bots failed running this test:
```
14:05:03 not ok 2529 parallel/test-http2-stream-destroy-event-order
14:05:03   ---
14:05:03   duration_ms: 120.35
14:05:03   severity: fail
14:05:03   exitcode: -15
14:05:03   stack: |-
14:05:03     timeout
14:05:03   ...
```
https://github.com/nodejs/node/blob/f5512ff61ecb668c2f49b7c05d3227ef7aa5e85f/test/parallel/test-http2-stream-destroy-event-order.js

But maybe that's a known problem? https://github.com/nodejs/node/pull/29889#issuecomment-543726136

vtjnash commented:
Bump???

Also it maybe also hit a more serious (unrelated) bug in nodejs? (https://ci.nodejs.org/job/node-test-commit-arm/28686/nodes=debian9-docker-armv7/)
```
not ok 2700 sequential/test-http2-ping-flood
  ---
  duration_ms: 3.457
  severity: crashed
  exitcode: -4
  stack: |-
  ...
```

cjihrig commented:
The failures are at least not completely related to your changes here.

test-http2-stream-destroy-event-order is being tracked as a flaky test in Node in https://github.com/nodejs/node/issues/31281.

test-http2-ping-flood is being tracked as a flaky test in Node in https://github.com/nodejs/node/issues/31107.

vtjnash commented:
OK, good to know those are probably unrelated (or at least, pre-existing). I think it should be good to land now (esp. now that 1.34.1 is branched).

Important note though: the merging-person should be aware there's a potential merge conflict with #2602: this adds a test checking for AIX being broken in the way that #2602 would fix. Thus whichever one lands first will imply a need to update and re-CI the other one.

vtjnash commented:
libuv CI: https://ci.nodejs.org/job/libuv-test-commit/1722/ ✅ 
node CI: https://ci.nodejs.org/view/libuv/job/libuv-in-node/123/ ⚠️

cjihrig commented:
FYI - this was reverted in https://github.com/libuv/libuv/pull/2968.
